Add support for multiple keyboards to Windows raw input driver

briankendall opened the pull request:
## Description

Adds support for multiple keyboards to the Windows raw input driver, similar to how it supports multiple mice. It repurposes the "Mouse Index" setting, renaming it to "Mouse/Keyboard Index", so that it also specifies which keyboard is associated with a port.

### Notes and possible discussion points:

- The "Mouse Index" setting is only changed in English. Other languages that include it need that label to be retranslated. I've marked them with a `TODO/FIXME`.
- This PR does not change the name of the setting stored in retroarch.cfg, as doing so would break compatibility with previous releases. I considered changing `config_load_file()` so that it first tries loading the old setting `input_playerN_mouse_index` and then a new one `input_playerN_mouse_keyboard_index`, but doing so leaves the old setting behind in retroarch.cfg
- The keyboard event callback still doesn't support multiple keyboards as it doesn't provide a parameter that could specify which device triggered the event. Consequently any emulator that supports multiple keyboards will have to poll the state of every key it cares about for each keyboard every cycle, which usually means individually polling the state of _every_ key. This strikes me as quite inefficient. In practice though it doesn't seem to be a performance bottleneck, at least not on my six year old gaming PC.

cmitu commented:
Would this patch couple the mouse/index into one setting ? It seems unnecessary and could cause issues for certain setups - i.e. one keyboard with 2 mouse like devices (spinner/trackballs).

briankendall commented:
> Would this patch couple the mouse/index into one setting ? It seems unnecessary and could cause issues for certain setups - i.e. one keyboard with 2 mouse like devices (spinner/trackballs).

@cmitu I'm not entirely sure what you're asking. But in case it answers your question, this PR adds a new purpose to the mouse index setting so that it can be used for ports whose type is keyboard as well. It shouldn't interfere with any existing functionality since a port can only be assigned one type of device at a time.

If that doesn't answer you're question, would you please clarify?

cmitu commented:
@briankendall Let's say I use a keyboard-type device for both P1 and P2 (like an WinIPAC) and also 2 trackballs input devices, which are detected as mice. Would I be able to specify - with your new patch - that P1 is Keyboard 1/Mouse 1 and P2 has the same Keyboard 1/Mouse2 as inputs ?

briankendall commented:
@cmitu Ah, I see the issue. If it is possible to assign both a keyboard and a mouse to a single player / port then repurposing mouse index will make that setup you described impossible.

In that case, what's a better solution? Should there be a new setting called "keyboard index"?

cmitu commented:
> In that case, what's a better solution? Should there be a new setting called "keyboard index"?

Yes, IMHO the keyboard index should be kept separate (defaulting to 0, for the first keyboard).

briankendall commented:
@cmitu Okay I can make that change. I think I have a pretty good idea of how to add a new setting, in that I can just follow the same patterns as used with "mouse index", but Is there anything I should keep in mind? And any other contributors / maintainers that should be part of this discussion?

briankendall commented:
@cmitu I've added a keyboard index setting. There's a few caveats to the changes I made:

- Similar to last time, there is no translation of "Keyboard index" for any language other than English. I've added a `TODO/FIXME` to msg_hash_us.h mentioning this.
- Nearly all of the items in a port's input settings have their own icon for the xmb and ozone themes. There was no keyboard icon I could use so I assigned "keyboard index" to use the default "subsetting" icon. I've also added a `TODO/FIXME` for both xmb.c and ozone_texture.c for this.

briankendall commented:
Another issue I've noticed when testing this PR: RetroArch will only respond to keyboard input from the keyboard configured on port 1. Ordinarily it can be controlled by all connected keyboards. This means that it's possible for me to use the keyboard to change port 1's keyboard index, thus causing RetroArch to no longer be configurable via the keyboard. That strikes me as an important usability issue! Perhaps an additional change needs to be made so that it still uses all connected keyboards?

cmitu commented:
@briankendall do you have `menu_unified_controls` set in your configuration ? 
I'm not sure if it affects the behavior you're seeing, but looks similar to how gamepads are behaving w.r.t. controlling RetroArch's menus/hotkeys (i.e. only P1 gamepad has the control).
See https://github.com/libretro/RetroArch/issues/10032 or https://github.com/libretro/RetroArch/issues/3337 ? 

If it's not what you meant, then please ignore my comment.

briankendall commented:
@cmitu I do have that setting enabled, but it doesn't affect keyboard input. I've found the part of RetroArch's code where the keyboard state is queried for the UI, and it's hardcoded to only get the state of the keyboard on port index 0 (i.e. port 1). Up until now this was fine as querying keyboard input state ignored the port parameter. But with this PR it's possible to configure port 0 to have a keyboard index that doesn't correspond to any actual keyboard.

I'm not sure what the best solution to this would be. Querying the keyboard state for all ports isn't a good solution because it's possible for all the ports to be configured to a keyboard index that doesn't exist. What strikes me as the correct solution is being able to query the state of a key in a way that explicitly ignores the port, so that it'll return true if any keyboard has that key down. But that would require a change to libretro's API.

Another possible solution would be for RetroArch's UI to use a keyboard callback instead of querying the input driver for key state, since those callbacks aren't triggered by the current input driver. But that would be a pretty significant change and I don't think I'm nearly familiar enough with the codebase to pull it off without introducing bugs!

dannyrb commented:
👀  I have been watching this issue. I have two "keyboard input devices" that I need to be able to bind separately. I would be willing to pay a small bug bounty to help this issue get across the finish line. I'm also a C#/JS dev, and could lend some support in kind if additional effort is needed to complete this.

briankendall commented:
I think what this PR needs at this point is for one of the RetroArch devs to comment on the issue I raised about being able to change the keyboard on port 1 to a non-existent keyboard index which breaks keyboard control of RetroArch's UI.

Thinking about this again after nine months, maybe the simplest solution is to force port 1 to always use keyboard index 0? As long as there's a keyboard connected to the system, that will match one of them.

Once that's done, I could fix the new conflicts that have emerged since March and fix the broken checks. Then this could get merged, provided the RetroArch devs don't raise any other concerns or require additional changes.

However, even when it's merged, it's still up to the various cores to make use of the ability to recognize multiple keyboards. I made a fork of the Flycast core that got multiple keyboards working in Typing of the Dead, but there's no point in trying to get those changes merged in until this PR is merged.

dannyrb commented:
> However, even when it's merged, it's still up to the various cores to make use of the ability to recognize multiple keyboards. I made a fork of the Flycast core that got multiple keyboards working in Typing of the Dead, but there's no point in trying to get those changes merged in until this PR is merged.

Ah. That's interesting. I had incorrectly assumed that being able to detect and bind on the device would resolve the issue. I wasn't aware that cores would need to be updated. Is there more information floating around on how cores consume the configured bindings?

I guess I'm curious about how these different things interact:
- Device(s)
- Drivers
- Bindings
- Libretro
- Cores

After this is merged, I would be interested in helping existing cores update to use the new functionality.

> Thinking about this again after nine months, maybe the simplest solution is to force port 1 to always use keyboard index 0? As long as there's a keyboard connected to the system, that will match one of them.

It's my understanding that you can detect which devices are keyboards, which devices are mice, and their hardware IDs. However, the device's numeric id (or its dynamic id) is not always consistent. It's worth noting that exact models/versions of software would have duplicate Hardware IDs, so the most reliable "unique" is the dynamic device id (really just it's index in an array?).

So, I think your suggestion would work. Find the first keyboard in the list, and use it. The biggest issue being that this may not always be the same keyboard. You could also allow keyboard input from all keyboards for the menu, but then use configured bindings for Cores?

All of my knowledge is coming from trying to resolve this at the driver level. I've been toying with the [Intereception Driver](https://github.com/oblitum/Interception) to fix my multiple keyboard issue at the driver level. Unfortunately, if you use the driver, RetroArch does not pick up intercepted keys (intercept A, send Z). 

----

I know it's not much, but I could probably add $250 to the pot for a solution that helps resolve these issues. As well as donate programming time; although I may not be very useful without more context/info.

briankendall commented:
> Ah. That's interesting. I had incorrectly assumed that being able to detect and bind on the device would resolve the issue. I wasn't aware that cores would need to be updated. Is there more information floating around on how cores consume the configured bindings?

I've only investigated one libretro core so far (Flycast) and what I found is that they used the libretro API that allows you to set up a callback function that triggers whenever the user presses a key on the keyboard. This callback functionality does not use the input driver, so in the case of keyboards it provides no information concerning what keyboard triggered the key.

The alternative, and this is the change I made to my personal fork of Flycast, is to use a different libretro API that polls the state of each keyboard on each port. (It's less efficient, but in practice there was no performance degradation on my dated gaming rig.)

However since the callback method is easier and faster, I'm guessing that most libretro cores that need access to the keyboard would opt to use that instead of the input driver. But that will vary on a core-by-core basis.

> It's my understanding that you can detect which devices are keyboards, which devices are mice, and their hardware IDs.  ...

I think this is getting into issues that are already solved in the changes I made to the Windows rawinput driver. It already finds all of the connected keyboards and assigns them an index in the order that they were detected, ignoring other irrelevant device types. The indices are contiguous (so the first keyboard has index 0, the second keyboard has index 1, etc.).

The specific problem that needs to be solved is that I added a new setting for each controller port named "keyboard index" that specifies which keyboard should be assigned to that port, as indexed by the driver. (This is analogous to the "mouse index" setting.) If the user changes the "keyboard index" setting for port 1 to an invalid index, suddenly it's impossible to use the keyboard to navigate RetroArch's GUI because it only listens for key events on whichever keyboard is configured for port 1.

markwkidd commented:
@jdgleaver I hope it's OK to tag you in this PR. it doesn't look like it ever got a code review, and based on recent discussions, maybe this is in your area of interest

ofry commented:
@briankendall There are merge conflicts. Could you fix them?

briankendall commented:
> @briankendall There are merge conflicts. Could you fix them?

Yes, I'll try to get to it in the not too distant future.

If I'm going to work on this PR again, I'd like to try and settle the issue regarding Retroarch's GUI that I brought up. Unless anyone has any objections, I'm going to see if I can set it up so that the keyboard on port 1 can't have its keyboard index changed.

gtxaspec commented:
That sounds good, waiting to play typing of the dead using this.

ofry commented:
@briankendall There is October now, no progress :(

briankendall commented:
> @briankendall There is October now, no progress :(

Sorry everybody, I've been so busy with other things in my life (including other projects) that I haven't had time to turn my attention to this. That said, I do still want to play multiplayer Typing of the Dead with my friends, so I do want to continue working on it.
